[daronoh] iP

daronoh opened the pull request:
### DGPT
> “Your mind is for having ideas, not holding them.” – David Allen

DGPT is a chatbot meant to help you with your schedule. It can

- keep track of tasks `ToDo`
- keep track of tasks with a `Deadline`
- keep track of an `Event` that have a start and end time

All you need to do is,

1. download it from [here](https://github.com/daronoh/ip/releases/tag/A-Jar)
2. double click it
3. add your tasks
4. let it remember it for you 😄 

and best of all, it's ~$99.99~ **free**!

Features:

- [x] Managing tasks
- [x] Managing deadlines
- [x] Managing events
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

daronoh committed:
Renamed the chatbot, as well as greet and exit feature

daronoh committed:
Implement Echo functionality

daronoh committed:
implement list functionality

daronoh committed:
added task class as well as mark as done functionality

daronoh committed:
Added Todo, Events, and Deadline classes and functionality

daronoh committed:
Merge remote-tracking branch 'origin/add-gradle-support'

daronoh committed:
Added automated text UI testing

daronoh committed:
Added error handling using IncorrectInputException and TaskNotFoundException

daronoh committed:
Implemented a delete request to remove a task from the list

daronoh committed:
Fix the styling and javadoc

daronoh committed:
Add reading and writing tasks from hard disk

Cannot save tasks betweem uses of the bot.

Being able to save tasks between uses allows the
bot to achieve its intended functionality.

Let's create loadExistingFile and saveToFile
functions that handle the reading and writing
of tasks, and use them in our main programme.

daronoh committed:
Merge branch 'branch-Level-7'
Merge Branch-Level-7 to master

daronoh committed:
Modify Deadline and Event to use LocalDateTime

Deadline and Event uses String to store date and
time related information.

LocalDateTime provides better functionality, such
as formatting and error handling, over using
Strings.

Let's change Deadline and Event to make use of
LocalDateTime to store information instead of
Strings.

daronoh committed:
Merge branch 'branch-Level-8'
Merge branch-Level-8 to master branch

daronoh committed:
Add OOP principles to DGPT

DGPT does not utilise OOP principles.

extracting functionalities into their own classes
help with organising code as well as logic.

Let's create classes to handle the UI, reading
and writing to the hard drive, parsing, as well
as the task list.

daronoh committed:
Organise into packages

daronoh committed:
Organise into packages

daronoh committed:
Merge branch 'branch-Packages'
Merge branch-Packages

daronoh committed:
Add Gradle and Junit testing

No form of testing for DGPT.

Testing can help to ensure that the product works
as intended as we work on it.

Let's add Gradle and JUnit tests to help automate
the testing process.

daronoh committed:
Merge branch 'branch-JUnit'

daronoh committed:
Change settings for Jar file

daronoh committed:
Merge branch 'branch-Jar'

daronoh committed:
Implement a "find" command to search for keywords

daronoh committed:
Add JavaDoc comments to classes and methods

daronoh committed:
Enforce java coding style standards

daronoh committed:
Merge branch 'branch-A-JavaDoc'

daronoh committed:
Merge branch 'branch-A-CodingStandard'

daronoh committed:
Merge branch 'branch-Level-9'

Domokunx commented on src/main/java/dgpt/DGPT.java:
```diff
@@ -0,0 +1,87 @@
+ * It allows users to add tasks, mark them as done, unmark them, view the list of tasks, as well as find tasks using
+ * keywords. It is able to save and load existing data using the local hard drive.
+ */
+public class DGPT {
```
Abbreviations should not be fully capitalized, unless they are well known or easily recognized

Domokunx commented on src/main/java/dgpt/DGPT.java:
```diff
@@ -0,0 +1,87 @@
+
+import java.io.IOException;
+import java.util.Scanner;
+
```
Nice organization of import statements!

Domokunx commented on src/main/java/dgpt/DGPT.java:
```diff
@@ -0,0 +1,87 @@
+
+
+    /**
+     * Main method to start the program and interact with the user.
```
```suggestion
     * Starts the program and interacts with the user.
```

Domokunx commented on src/main/java/dgpt/Parser.java:
```diff
@@ -0,0 +1,131 @@
+     * @throws IncorrectInputException If the input format is incorrect or missing required parameters.
+     * @throws TaskNotFoundException If the command refers to a task that cannot be found.
+     */
+
```
Empty line could be removed?

Domokunx commented on src/main/java/dgpt/Parser.java:
```diff
@@ -0,0 +1,131 @@
+     */
+
+    public static void parse(String input, TaskList taskList, Ui ui) throws IncorrectInputException,
+            TaskNotFoundException {
```
IncorrectInputException and TaskNotFoundException sound like RunTimeExceptions, so maybe you don't have to throw them specifically?

Domokunx commented on src/main/java/dgpt/Parser.java:
```diff
@@ -0,0 +1,131 @@
+    public static void parse(String input, TaskList taskList, Ui ui) throws IncorrectInputException,
+            TaskNotFoundException {
+
+        String[] command = input.split(" ", 2);
```
Collections should be plural, so maybe inputs?

Domokunx commented on src/main/java/dgpt/Storage.java:
```diff
@@ -0,0 +1,142 @@
+package dgpt;
+
+import dgpt.exception.DgptFileNotFoundException;
+import dgpt.task.*;
```
wildcard imports are not recommended

Domokunx commented on src/main/java/dgpt/Storage.java:
```diff
@@ -0,0 +1,142 @@
+    public List<Task> load() throws IOException, DgptFileNotFoundException {
+        List<Task> res = new ArrayList<>();
+        try {
+
```
empty line could be removed

Domokunx commented on src/main/java/dgpt/Ui.java:
```diff
@@ -0,0 +1,169 @@
+ * The Ui class is responsible for handling the user interface output for the DGPT application.
+ * It contains methods to display various messages and information to the user.
+ */
+
```
empty line could be removed

Domokunx commented on src/main/java/dgpt/Ui.java:
```diff
@@ -0,0 +1,169 @@
+            System.out.println(index + "." + t.toString());
+            index++;
+        }
+        System.out.println("-----------------------");
```
Maybe consider refactoring this divider print into a helper function? 
Easier to modify in the future if you wish to maybe allow style changes or to adjust to the console width.

Domokunx commented on src/main/java/dgpt/exception/DgptFileNotFoundException.java:
```diff
@@ -0,0 +1,12 @@
+ * Thrown to indicate that a specific file was not found.
+ * This exception is used by the {@code Storage} class when it cannot locate the file specified.
+ */
+public class DgptFileNotFoundException extends Exception {
```
I feel like this should extend RuntimeException instead?

Domokunx commented on src/main/java/dgpt/exception/IncorrectInputException.java:
```diff
@@ -0,0 +1,11 @@
+/**
+ * Thrown to indicate that the number of inputs after the command is incorrect.
+ */
+public class IncorrectInputException extends Exception {
```
This also feels more like a RuntimeException.

Domokunx commented on src/main/java/dgpt/exception/TaskNotFoundException.java:
```diff
@@ -0,0 +1,11 @@
+/**
+ * Thrown to indicate that the command is not recognised.
+ */
+public class TaskNotFoundException extends Exception {
```
This also feels like a RuntimeException

Domokunx reviewed (commented):
Looks great! Easy to read with well written documentation. 😄 
Just some cosmetic nitpicks to consider!

Also consider having your custom exceptions inherit from RuntimeException instead, because they don't feel like checked exceptions the compiler can check without executing the program.

Professional-Procrastinat0r commented on src/main/java/dgpt/Parser.java:
```diff
@@ -0,0 +1,131 @@
+        String[] command = input.split(" ", 2);
+
+        switch (command[0]) {
+        case "list" -> {
```
Adheres to coding convention for `case` clauses not having indentations

Professional-Procrastinat0r commented on src/main/java/dgpt/Ui.java:
```diff
@@ -0,0 +1,169 @@
+    /**
+     * Displays a welcome message to the user.
+     */
+    public void showWelcome() {
```
Adheres to using verbs for methods

Professional-Procrastinat0r commented on src/main/java/dgpt/Ui.java:
```diff
@@ -0,0 +1,169 @@
+ * It contains methods to display various messages and information to the user.
+ */
+
+public class Ui {
```
Adheres to using nouns for class names

Professional-Procrastinat0r commented on src/main/java/dgpt/task/Task.java:
```diff
@@ -0,0 +1,80 @@
+abstract public class Task {
+
+    protected String description;
+    protected boolean isDone;
```
Good use of boolean-sounding name for boolean variable

Professional-Procrastinat0r commented on src/main/java/dgpt/task/TaskList.java:
```diff
@@ -0,0 +1,146 @@
+     * @param keyword The keyword to search for in the task descriptions.
+     * @return A list of tasks that contain the keyword in their description.
+     */
+    public List<Task> findTasks(String keyword) {
```
Follows naming convention for methods, but can consider adding more description? Eg,`findTasksByDescription`

daronoh committed:
Add CheckStyle to project

daronoh committed:
Merge branch 'branch-A-CheckStyle'

daronoh committed:
Utilise FXML to add a GUI element

Text-based UI is not user friendly.

Having a GUI will make user experience better.

Implement FXML to add GUI functionality.

Let's create FXML controllers and views.

daronoh committed:
Merge branch 'branch-Level-10'

Xczheng0105 commented on src/main/java/dgpt/Dgpt.java:
```diff
@@ -0,0 +1,51 @@
+package dgpt;
```
Good organisation of import statements.

Xczheng0105 commented on src/main/java/dgpt/Parser.java:
```diff
@@ -0,0 +1,169 @@
+                output.append(e.getMessage()).append("\n");
+            }
+        }
+        default -> {
```
Good inclusion of default case for switch-case block.

Xczheng0105 commented on src/main/java/dgpt/Storage.java:
```diff
@@ -0,0 +1,148 @@
+import dgpt.task.Task;
+import dgpt.task.TaskList;
+import dgpt.task.ToDo;
+
```
Do remove these empty lines.

Xczheng0105 commented on src/main/java/dgpt/task/Event.java:
```diff
@@ -0,0 +1,50 @@
+
+    @Override
+    public String toString() {
+        return "[E]" + super.toString() + " (from: " + this.getFromTime() + " to: "
```
Good use of spacing for operators. Conforms to coding standard.

Xczheng0105 reviewed (commented):
Conforms to the coding standard generally. LGTM =)

daronoh committed:
Add assertions
No assertions are used.

Assertions can help to verify if certain assumptions are valid in our
program.

Add assertions to ensure this functionality is present.

daronoh committed:
Improve code quality
Readability of the code can be improved.

Improving the code quality can improve the
readability of the code, as well as help with
finding bugs and optimization down the road.

Reduce repeated code and arrowhead style code.

Let's refactor certain parts of the code to
improve code quality.

daronoh committed:
Merge pull request #1 from daronoh/branch-A-Assertions

Add assertions

daronoh committed:
Merge pull request #2 from daronoh/branch-A-CodeQuality

Improve code quality

daronoh committed:
Set up continuous integration

daronoh committed:
Enable assertions in gradle

daronoh committed:
Remove text-ui-test functionality

daronoh committed:
Remove IO redirection tests

daronoh committed:
Correct Checkstyle errors

daronoh committed:
Add a new type of task which is recurring
Some tasks can be recurring and Dgpt is unable to
differentiate recurring tasks from those
that are not.

Having this functionality will improve the
effectiveness of this product.

Add a recurring type of task, with its respective
formatted command.

daronoh committed:
Merge branch 'branch-BCD-Extension'

daronoh committed:
Improve code quality of Parse

Parse function in Parser is not adhering to SLAP.

Adhering to SLAP will allow our code to be more
readable.

Abstract certain functionalities out into their
own functions.

Let's create separate functions to handle certain
asepcts of the Parse function.

daronoh committed:
Improve code quality of Parser class

Parse function in Parser does not apply SLAP when
the "delete" command is found.

Adhering to SLAP will make the code more readable.

Let's abstract out the functionality of delete
into a parseDelete function.

daronoh committed:
Further improve code quality of Parser class

Parse function in parser class still does not
adhere to SLAP.

adhering to SLAP will make the code more readable.

Let's abstract out the functionality for "find",
"save" and "recurring" commands into their own
functions.

daronoh committed:
Apply SLAP principle to Parser class

Parse function in Parser class does not adhere to SLAP principle.

Adhering to SLAP will make the code more readable.

Let's abstract out all further levels of functionality to separate
functions, leaving parse to handle a single level of abstraction.

daronoh committed:
Improve error handling

Parsing an integer in "mark", "unmark" and "delete" command does
not account for the case where the input is of an invalid format.

Handling this error can improve user experience as they will get
a feedback when this error occurs.

Let's catch this error and print it out using the Ui.

daronoh committed:
Apply SLAP to Storage class

Save function in Storage class does not adhere to SLAP.

Adhering to SLAP will ensure that the code is more readable.

Let's abstract out the functionalities from the save function into
their own separate functions.

daronoh committed:
Update readme

DGPT does not have a user guide.

Having a user guide will provide users with a brief introduction
to the product and its functionalities.

Let's update the readme to include a user guide for the product.

daronoh committed:
Bug fix EventTest class

Expected outputs for the EventTest class are incorrect.

The expected output should have a 0 in front of the date if the
date is only 1 digit.

Let's add a 0 in front of the expected output so that the test will
work as intended.

daronoh committed:
Bug fix for date-time parsing

No standard locale and timezone set.

Setting a standard locale and timezone will help users on different
systems and locations experience similar results.

Let's implement a standard locale and timezone.

daronoh committed:
Standardize output for time format

Output of time formats are dependent on the local machine running.

The format of the output should be consistent.

Let's enforce AM and PM outputs to be caps.

daronoh committed:
Improve error handling

The UI was not showing informative error messages.

Showing clearer error messages can improve user experience.

Let's have the bot give a more human-like error response.

daronoh committed:
Further improve error handling

Error message shown doesn't feel very human-like.

Providing a more human-like error message can improve the user
experience as the product is a chatbot.

Let's alter the sentence structure of the response such that it is
more speech-like.

daronoh committed:
Bug fix typo errors

There are typo errors in the error messages shown.

Typo errors in the error messages can mislead users into thinking
that their error was caused by something else.

Let's amend the typos.
